android/src/main/java/app/edge/reactnative/core/EdgeNative.java

```diff
@@ -97,6 +110,87 @@ private void handleCall(
     }
   }
 
+  private void handleFetch(@NonNull JSONArray args, @NonNull PendingCall promise)
+      throws JSONException {
+    String uri = args.getString(0);
+    String method = args.getString(1);
+    JSONObject headers = args.getJSONObject(2);
+    String body = args.optString(3);
+    boolean bodyIsBase64 = args.optBoolean(4);
+
+    HttpURLConnection connection = null;
+    try {
+      // Set up the HTTP connection:
+      connection = (HttpURLConnection) new URL(uri).openConnection();
+      connection.setRequestMethod(method);
+      connection.setDoInput(true);
+      connection.setUseCaches(false);
+
+      // HTTPS needs help:
+      if (connection instanceof HttpsURLConnection) {
+        HttpsURLConnection httpsConnection = (HttpsURLConnection) connection;
+        SSLContext context = SSLContext.getInstance("TLS");
+        context.init(null, null, null);
+        httpsConnection.setSSLSocketFactory(context.getSocketFactory());
+      }
+
+      // Add the headers:
+      for (Iterator<String> it = headers.keys(); it.hasNext(); ) {
+        String key = it.next();
+        connection.setRequestProperty(key, headers.getString(key));
+      }
+
+      // Add the body:
+      if (body.length() > 0) {
+        byte[] bodyData =
+            bodyIsBase64
+                ? Base64.decode(body, Base64.DEFAULT)
+                : body.getBytes(StandardCharsets.UTF_8);
+        connection.setRequestProperty("Content-Length", Integer.toString(bodyData.length));
+        connection.setDoOutput(true);
+        OutputStream outStream = connection.getOutputStream();
+        outStream.write(bodyData);
+        outStream.flush();
+        outStream.close();
+      }
+
+      // Make the request:
+      connection.connect();
+
+      // Read the response status:
+      JSONObject response = new JSONObject();
+      response.put("status", connection.getResponseCode());
+
+      // Read the response headers:
+      JSONObject responseHeaders = new JSONObject();
+      for (int i = 0; ; ++i) {
+        @Nullable String key = connection.getHeaderFieldKey(i);
+        @Nullable String value = connection.getHeaderField(i);
+        if (key == null && i == 0) continue; // Might be the HTTP status
+        if (key == null || value == null) break;
+        responseHeaders.put(key, value);
+      }
+      response.put("headers", responseHeaders);
+
+      // Read the response body:
+      StreamStringReader responseBody = new StreamStringReader();
+      responseBody.read(connection.getInputStream(), responseHeaders.optInt("Content-Length"));
+      try {
+        response.put("body", responseBody.getUtf8());
+        response.put("bodyIsBase64", false);
```

Comment:
The goal here is to send the response up to JS as efficiently as possible - whether we use utf8 or base64 for that doesn't matter. As long as we communicate which one we are using, the bytes will arrive on the JavaScript side without confusion.

Once we get the bytes into JavaScript, we can do whatever we want with them, including converting them from utf8 to something else. I agree that parsing the "Content-Type" header would make sense, but this would take a good amount of code, plus support for various character sets. We can add this later in JS if we truly need it, but realistically everybody uses utf8 these days.